connectors: Add Unittests for opaconnector

rohithdv opened the pull request:
This PR addresses #124 . 

This PR does the following changes to opa connector:
(1) Minor redesign of opa connector to make it suitable for unit tests
(2) Preliminary unit tests for opa connector

To run the unit test, execute the following:
(1) cd to connectors/opa
(2)In the Makefile in this dir, set/unset NEED_TEST_COVERAGE_STATISTICS and NEED_TEST_COVERAGE_STATISTICS_IN_HTML as per requirement. 
(3) Execute make test 

Note: This PR is a revised version of PR https://github.com/IBM/the-mesh-for-data/pull/47 which was closed due to some errors due to DCO changes.

shlomitk1 commented on connectors/opa/lib/evaluate_opa.go:
```diff
@@ -57,17 +57,17 @@ func EvaluateExtendedPoliciesOnInput(inputMap map[string]interface{}, opaServerU
+	// data, _ := ioutil.ReadAll(res.Body)
+	// fmt.Printf("body from policy http response: %s\n", data)
+	// fmt.Printf("status from policy http response: %d\n", res.StatusCode)
+	// res.Body.Close()
```
Can the commented code be removed?

rohithdv replied:
fixed. Thanks.

roee88 commented on connectors/opa/testutil/helper.go:
```diff
@@ -0,0 +1,192 @@
+	"google.golang.org/grpc"
+)
+
+var EnvValues = map[string]string{
+	"CONNECTION_TIMEOUT":    "120",
+	"CATALOG_CONNECTOR_URL": "localhost:50084",
+	"OPA_SERVER_URL":        "localhost:8282",
+}
+
+func GetEnvironment() (int, string, string) {
+	timeOutInSecs := EnvValues["CONNECTION_TIMEOUT"]
+	timeOutSecs, _ := strconv.Atoi(timeOutInSecs)
+
+	catalogConnectorURL := EnvValues["CATALOG_CONNECTOR_URL"]
+	opaServerUrl := EnvValues["OPA_SERVER_URL"]
+	log.Printf("EnvVariables = %v\n", EnvValues)
+
+	return timeOutSecs, catalogConnectorURL, opaServerUrl
+}
```
Why not just use constants directly?

rohithdv replied:
having the EnvValues variable gives us flexibility to change these env variables in different testcases (if need be) as shown in example [here](https://github.com/rohithdv/the-mesh-for-data/blob/6bfa712c3ac2450a9f9eba2d4051384f9141b514/connectors/opa/lib/opa_reader_test.go#L43).

roee88 replied:
Seems too complex but I assume it's a preparation for actually taking these values from environment variables.

rohithdv replied:
actually we wanted to not depend on env variables for running the test cases. hence we created this struct.

roee88 commented on connectors/opa/testutil/helper.go:
```diff
@@ -0,0 +1,192 @@
+		if purpose == "marketing" {
+			newEnforcementAction := ConstructRedactColumn("nameDest")
+			enforcementActions = append(enforcementActions, newEnforcementAction)
+			newUsedPolicy = &pb.Policy{Description: "reduct columns with name nameOrig and nameDest  in datasets with Finance"}
```
finance or marketing?

rohithdv replied:
we have been using 2 purposes by convention - fraud-detection and marketing in [other test cases](https://github.com/rohithdv/the-mesh-for-data/blob/6bfa712c3ac2450a9f9eba2d4051384f9141b514/pkg/policy-compiler/pc-bl/policy_compiler_test.go#L62). so we have used "marketing" here since in the opa connector test case context any purpose would do. these are just for testing purposes.

roee88 replied:
This is within a block of an if statement that compares to "marketing"

rohithdv replied:
Actually what the description means is “redact columns with name nameOrig and nameDest  in datasets which have been tagged with Finance” ..The purpose could be marketing or fraud-detection or any other purpose..I have edited the description to the above to avoid confusion. Thanks.

roee88 reviewed (commented):
Added some comments. I'm still struggling with `testutil/helper.go`. It includes too much and actually having it hides what the test does (looking at the test file all you can tell is we send a request and expect some results, go to helper to see what)

froesef commented on connectors/opa/lib/opa_reader_test.go:
```diff
@@ -0,0 +1,49 @@
+
+	srv := NewOpaReader(opaServerUrl)
+	catalogReader := NewCatalogReader(catalogConnectorURL, timeOutSecs)
+	policiesDecisions, _ := srv.GetOPADecisions(applicationContext, catalogReader)
```
You should check that the error is nil:
```
policiesDecisions, err := srv.GetOPADecisions(applicationContext, catalogReader)
assert.NilError(t, err)
```

rohithdv replied:
Fixed by using the assert function from gotest.tools/assert. Thanks.

froesef commented on connectors/opa/main.go:
```diff
@@ -0,0 +1,93 @@
+		log.Println("GetOPADecisions err:", err)
+		return nil, err
+	}
+	jsonOutput, _ := json.MarshalIndent(eval, "", "\t")
```
Please handle the error here. Thus if there is an error you probably should not log the jsonOutput string.

rohithdv replied:
Fixed. Thanks.

rohithdv commented on connectors/opa/lib/evaluate_opa.go:
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
-package main
+package opaconnbl
```
sorry I didn't understand this comment. is it something to do with the naming convention of the package ?

roee88 replied:
I assumed this is a leftover and you wanted it to match the directory name.

rohithdv replied:
yes right. Sorry I got confused with the older dir structure. I have changed the package name to lib now. Thanks.

rohithdv commented on connectors/opa/lib/opa_reader.go:
```diff
@@ -1,79 +1,31 @@
 // SPDX-License-Identifier: Apache-2.0
 
-package main
+package opaconnbl
```
sorry I didn't understand this comment. is it something to do with the naming convention of the package ?

rohithdv replied:
I have changed the package name to lib now. Thanks.

rohithdv commented on connectors/opa/lib/catalog_reader.go:
```diff
@@ -0,0 +1,99 @@
+	metadataMap := make(map[string]interface{})
+	err = json.Unmarshal(responseBytes, &metadataMap)
+	if err != nil {
+		log.Printf("error in unmarshalling responseBytes: %v", err)
```
fixed. Thanks.

rohithdv commented on connectors/opa/lib/catalog_reader.go:
```diff
@@ -0,0 +1,99 @@
+		}
+		return nil, fmt.Errorf("error sending data to External Catalog Connector (datasetID = %s): %v", datasetID, err)
+	}
+	log.Println("***************************************************************")
```
fixed. Thanks.

rohithdv commented on connectors/opa/lib/catalog_reader.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func (c *CatalogReader) GetDatasetMetadata(ctx *context.Context, client pb.DataCatalogServiceClient, datasetID string, appID string) (map[string]interface{}, error) {
+	objToSend := &pb.CatalogDatasetRequest{AppId: appID, DatasetId: datasetID}
```
In the WKC context, all the connectors need the appID to retrieve the credentials for accessing WKC as shown [here](https://github.ibm.com/data-mesh-research/WKC-connector/blob/43100b73fb150c7acac7ac8dad8304f5fa0568d1/java/src/main/java/com/ibm/wkc/WKCCredentials.java#L69). 

In the open source context , I dont think we are using appID for retrieving credentials as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/vault/vault_connector.go#L32). Also we are NOT using appID to retrieve catalog information from the egeria catalog connector as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/egeria/src/main/java/com/ibm/egeria/EgeriaClient.java#L377). For Policy manager connector (OPA Connector) also we are NOT using appID as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/opa/opa_connector.go#L51). 

To summarise, we are using appID only in the context of WKC. We are not using appID when we are using open source components like opa connector, egeria connector and vault connector.

rohithdv commented on connectors/opa/Makefile:
```diff
@@ -7,9 +7,11 @@ include $(ROOT_DIR)/.mk/docker.mk
 
 OPA_PATH=${ABSTOOLBIN}/opa
 OPA_DEFAULT_POLICY_FILE=${ROOT_DIR}/third_party/opa/opa-policy.rego
+NEED_TEST_COVERAGE_STATISTICS=1
+NEED_TEST_COVERAGE_STATISTICS_IN_HTML=0
```
@hunchback Can you please comment on this? Thanks.

rohithdv commented on connectors/opa/lib/catalog_reader.go:
```diff
@@ -0,0 +1,99 @@
+	"google.golang.org/grpc/status"
+)
+
+// Reader from catalog, single instance for the connector, not dependent on the request
```
We have separated the OPA connector functionality into 2 structs - CatalogReader (which has functions which reads info from the catalog) and OPAReader (which uses the catalog metadata and performs policy evaluation). 

CatalogReader is a struct which has information to read from catalog. This struct does not have any information related to the any specific application context. Any request specific info is passed as parameters to functions belonging to this struct. I have updated the comment of CatalogReader based on this explanation. Hope it is ok. 

I think both these structs are part of OPAConnector and they are just separated based on their distinct functionalities. In the earlier version of OpaConneector code, we had merged the functions related to catalog reader and opa reader which was a bit confusing. Now there are separated for better clarity.

roee88 replied:
What makes catalog reader specific to the opa connector. It looks like a generic client to a catalog connector

rohithdv replied:
ok thanks.

rohithdv commented on connectors/opa/lib/catalog_reader.go:
```diff
@@ -0,0 +1,99 @@
+	"google.golang.org/grpc/status"
+)
+
+// Reader from catalog, single instance for the connector, not dependent on the request
```
fixed. Please check if it is ok. Thanks.

rohithdv commented:
> Added some comments. I'm still struggling with `testutil/helper.go`. It includes too much and actually having it hides what the test does (looking at the test file all you can tell is we send a request and expect some results, go to helper to see what)

basically testutil/helper.go has functions to (1) to start mock catalog connector (2) to start mock opa server (3) implements GetDatasetInfo on mock catalog connector (4) function to return mock OPA server response (5) have GetExpectedOpaDecisions() which returns pb.PoliciesDecisions object which has OPA decisions that are expected based on what are configured in the mocks.  (6) additional helper functions which help in implementing (1) - (5) above. 

At a high level, testutil/helper.go is designed to help us to configure different test cases based on different scenarios which may arise when working with opa connector.
 
Please let me know if this is ok. Thanks.

roee88 reviewed (approved):
Replied to your comments, let me know if you want to address them before merging.

rohithdv commented:
> Replied to your comments, let me know if you want to address them before merging.

Addressed most of your comments. pls let me know if I need to do anything about this comment. https://github.com/IBM/the-mesh-for-data/pull/125#discussion_r529367702. Thanks.

rohithdv commented:
> > Replied to your comments, let me know if you want to address them before merging.
> 
> Addressed most of your comments. pls let me know if I need to do anything about this comment. [#125 (comment)](https://github.com/IBM/the-mesh-for-data/pull/125#discussion_r529367702). Thanks.

Actually what the description means is “redact columns with name nameOrig and nameDest  in datasets which have been tagged with Finance” ..The purpose could be marketing or fraud-detection or any other purpose..I have edited the description to the above to avoid confusion. Thanks.
